net/tcp/impl/posix.h

```diff
@@ -192,107 +181,92 @@ class Connection : public SocketHandle {
+      T* output_buffer, size_t max_length, BlockingReadPolicy policy = BlockingReadPolicy::ReturnASAP) {
+    BRICKS_NET_LOG("S%05d BlockingRead() ...\n", static_cast<SOCKET>(socket));
+    uint8_t* buffer = reinterpret_cast<uint8_t*>(output_buffer);
+    uint8_t* ptr = buffer;
+    size_t remaining_bytes_to_read;
+    while ((remaining_bytes_to_read = (max_length - (ptr - buffer))) > 0) {
+      BRICKS_NET_LOG("S%05d BlockingRead() ... attempting to read %d bytes.\n",
+                     static_cast<SOCKET>(socket),
+                     static_cast<int>(remaining_bytes_to_read));
 #ifndef BRICKS_WINDOWS
-      const ssize_t retval = ::recv(socket, raw_ptr, max_length_in_bytes - (raw_ptr - raw_buffer), 0);
+      const ssize_t retval = ::recv(socket, ptr, remaining_bytes_to_read, 0);
 #else
       const int retval =
-          ::recv(socket, reinterpret_cast<char*>(raw_ptr), max_length_in_bytes - (raw_ptr - raw_buffer), 0);
+          ::recv(socket, reinterpret_cast<char*>(ptr), static_cast<int>(remaining_bytes_to_read), 0);
 #endif
-      if (retval < 0) {
-        // TODO(dkorolev): Unit-test this logic.
-        // I could not find a simple way to reproduce it for the test. -- D.K.
+      BRICKS_NET_LOG(
+          "S%05d BlockingRead() ... retval = %d.\n", static_cast<SOCKET>(socket), static_cast<int>(retval));
+      if (retval > 0) {
+        ptr += retval;
+        if (policy == BlockingReadPolicy::ReturnASAP) {
+          return (ptr - buffer);
+        }
+      } else {
         // LCOV_EXCL_START
-        if (errno == EAGAIN) {
-          continue;
-        } else if (errno == ECONNRESET) {
-          // Allow one "Connection reset by peer" error to happen.
-          // Tested that this makes load test pass 1000/1000, while w/o this
-          // the test fails on some iteration with ECONNRESET on Ubuntu 12.04 -- D.K.
-          if (alive) {
-            alive = false;
-            continue;
+        if (retval < 0) {
+          if (policy == BlockingReadPolicy::ReturnASAP) {
+            return (ptr - buffer);
           } else {
-            if ((raw_ptr - raw_buffer) % sizeof(T)) {
-              BRICKS_THROW(SocketReadMultibyteRecordEndedPrematurelyException());
-            } else {
+            if (errno == EAGAIN) {
+              continue;
+            } else if (errno == ECONNRESET) {
+              BRICKS_NET_LOG("S%05d BlockingRead() : Connection reset by peer after reading %d bytes.\n",
+                             static_cast<SOCKET>(socket),
+                             static_cast<int>(ptr - buffer));
               BRICKS_THROW(ConnectionResetByPeer());
+            } else {
+              BRICKS_NET_LOG("S%05d BlockingRead() : Error after reading %d bytes, errno %d.\n",
+                             static_cast<SOCKET>(socket),
+                             static_cast<int>(ptr - buffer),
+                             errno);
+#ifdef BRICKS_WINDOWS
+              if (retval == -1) {
```

Comment:
`WSAGetLastError()` returns `WSAECONNRESET`, `Connection reset by peer`.

It is triggered rarely, and, interestingly, it is only by the `net/http`,
tests not the `net/tcp` or `net/api` ones.

I'll clean up this code wrt Windows vs. Linux `WSAGetLastError()` vs.
`errno`. Would prefer to do it once this pull request is merged though, to
avoid confusions.

Thanks,
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783


On Thu, Feb 26, 2015 at 1:36 AM, Alex Zolotarev <notifications@github.com>
wrote:

> In net/tcp/impl/posix.h
> <https://github.com/KnowSheet/Bricks/pull/94#discussion_r25414295>:
>
> >                BRICKS_THROW(ConnectionResetByPeer());
> > +            } else {
> > +              BRICKS_NET_LOG("S%05d BlockingRead() : Error after reading %d bytes, errno %d.\n",
> > +                             static_cast<SOCKET>(socket),
> > +                             static_cast<int>(ptr - buffer),
> > +                             errno);
> > +#ifdef BRICKS_WINDOWS
> > +              if (retval == -1) {
>
> WTF???
> What is returned by WSAGetLastError() call if retval == -1 (SOCKET_ERROR)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KnowSheet/Bricks/pull/94/files#r25414295>.
>